### PR TITLE
refactor(connlib): introduce `LoginUrl` component

### PR DESCRIPTION
**Pull request by thomaseizinger**
Currently, we are passing a lot of data into `Session::connect`. Half of this data is only needed to construct the URL we will use to connect to the portal. We can simplify this by extracting a dedicated `LoginUrl` component that captures and validates this data early.

Not only does this reduce the number of parameters we pass to `Session::connect`, it also reduces the number of failure cases we have to deal with in `Session::connect`. Any time the session fails, we have to call `onDisconnected` to inform the client. Thus, we should perform as much validation as we can early on. In other words, once `Session::connect` returns, the client should be able to expect that the tunnel is starting.

### CODE REVIEW

File: rust/connlib/clients/shared/src/lib.rs [hunk 1]
```diff
@@ -139,27 +133,18 @@ where
...
-                &callbacks
-            );
-
             // This is kinda hacky, the buffer size is 1 so that we make sure that we
             // process one message at a time, blocking if a previous message haven't been processed
             // to force queue ordering.
             let (control_plane_sender, mut control_plane_receiver) = tokio::sync::mpsc::channel(1);
 
-            let mut connection = PhoenixChannel::<_, IngressMessages, ReplyMessages, Messages>::new(Secret::new(SecureUrl::from_url(connect_url)), os_version_override, move |msg, reference, topic| {
+            let mut connection = PhoenixChannel::<_, IngressMessages, ReplyMessages, Messages>::new(Secret::new(SecureUrl::from_url(Url::parse(url.url().expose_secret().as_ref()).expect("we know it is a valid URL"))), os_version_override, move |msg, reference, topic| {
```
Comment (thomaseizinger) [inline rust/connlib/clients/shared/src/lib.rs:147]: This is a bit ugly. I am going to send another PR that cleans this up first and then rebase this one on top.

File: rust/connlib/clients/android/src/lib.rs [hunk 2]
```diff
@@ -411,11 +415,18 @@ fn connect(
...
-    let session = Session::connect(
+    let (private_key, public_key) = keypair();
+    let login = LoginUrl::client(
         api_url.as_str(),
-        secret,
+        &secret,
         device_id,
         Some(device_name),
+        public_key.to_bytes(),
+    )?;
```
Comment (thomaseizinger) [inline rust/connlib/clients/android/src/lib.rs:425]: With this PR, creating a new `Session` is slightly more verbose. In particular, one needs to explicitly create a keypair first. That is because `LoginUrl` lives in `phoenix-channel`. That is where it should live IMO because it is its sole consumer and the format defined by the URL is directly coupled to the use of `phoenix-channel`[^1].

However, I think `phoenix-channel` should _not_ depend on boringtun and key-types, thus you have to pass in 32-bytes which represent the public key.

Overall, I think this is an okay trade-off:

1. We are now unifying, how we construct `LoginUrl`s across clients, gateways and the relay.
2. The only fallible point of `Session` is now creation of the runtime. As we continue to chip away at simplifying session, `connect` might eventually become infallible.

[^1]: At a future point, we might even be able to tie the "login-topic" to the `LoginUrl` via the type system, i.e. you will automatically login as a client if you made a `LoginUrl::client`.

File: rust/connlib/shared/Cargo.toml [hunk 3]
```diff
@@ -42,6 +42,7 @@ libc = "0.2"
 dns-lookup = { workspace = true }
 known-folders = "1.1.0"
 snownet = { workspace = true }
+phoenix-channel = { workspace = true }
```
Comment (thomaseizinger) [inline rust/connlib/shared/Cargo.toml:44]: This dependency is not ideal but will go away once we delete this implementation of the phoenix-channel client.

File: rust/phoenix-channel/src/login_url.rs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn inner(&self) -> &Url {
+        &self.url
+    }
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn host(&self) -> &str {
+        &self.host
+    }
```
Comment (thomaseizinger) [inline rust/phoenix-channel/src/login_url.rs:131]: These will eventually be `pub(crate)` once we delete the other phoenix-channel client impl.

File: rust/relay/src/main.rs [hunk 5]
```diff
@@ -250,33 +250,21 @@ fn env_filter() -> EnvFilter {
...
-        url.query_pairs_mut()
-            .append_pair("ipv4", &public_ip4_addr.to_string());
-    }
-    if let Some(public_ip6_addr) = args.public_ip6_addr {
-        url.query_pairs_mut()
-            .append_pair("ipv6", &public_ip6_addr.to_string());
-    }
-    if let Some(name) = args.name.as_ref() {
-        url.query_pairs_mut().append_pair("name", name);
-    }
```
Comment (thomaseizinger) [inline rust/relay/src/main.rs:276]: This is now unified with clients and gateways.

File: rust/gui-client/src-tauri/src/client/deep_link/linux.rs [hunk 6]
```diff
@@ -13,7 +12,7 @@ impl Server {
         Ok(Self {})
     }
 
-    pub(crate) async fn accept(self) -> Result<Secret<SecureUrl>, Error> {
+    pub(crate) async fn accept(self) -> Result<SecretString, Error> {
```
Comment (thomaseizinger) [inline rust/gui-client/src-tauri/src/client/deep_link/linux.rs:15]: We can split this out into a different PR if we want. The `SecureUrl` type was originally only meant for the phoenix-channel so I replaced it with a `SecretString` here because I didn't want to keep the type just for this.

File: rust/connlib/clients/apple/src/lib.rs [hunk 7]
```diff
@@ -191,11 +193,19 @@ impl WrappedSession {
...
+        let (private_key, public_key) = keypair();
+        let login = LoginUrl::client(
             api_url.as_str(),
-            secret,
+            &secret,
             device_id,
             device_name_override,
+            public_key.to_bytes(),
+        )
+        .map_err(|e| e.to_string())?;
```
Comment (thomaseizinger) [inline rust/connlib/clients/apple/src/lib.rs:204]: Now that the `LoginUrl` type is separate, we might be able to further simplify this based on which client calls it.

File: rust/connlib/clients/android/src/lib.rs (hunk 2 above)
Comment (ReactorScram) [reply rust/connlib/clients/android/src/lib.rs:425]: I like that this opens a possibility for clients to persist their private key to disk if we ever want TOFU semantics with an even stronger "We can't decrypt your traffic even if we want" claim after 1.0

File: rust/connlib/clients/shared/src/lib.rs [hunk 8]
```diff
@@ -1,22 +1,21 @@
 //! Main connlib library for clients.
 pub use connlib_shared::messages::ResourceDescription;
-pub use connlib_shared::{Callbacks, Error};
+pub use connlib_shared::{keypair, Callbacks, Error, LoginUrl, LoginUrlError};
 pub use tracing_appender::non_blocking::WorkerGuard;
 
 use backoff::{backoff::Backoff, ExponentialBackoffBuilder};
-use connlib_shared::control::SecureUrl;
-use connlib_shared::{control::PhoenixChannel, login_url, CallbackErrorFacade, Mode, Result};
+use connlib_shared::StaticSecret;
```
Comment (ReactorScram) [inline rust/connlib/clients/shared/src/lib.rs:7]: I saw the Dalek docs said it's better to use EphemeralSecret. I assume the reason we use StaticSecret is so we can share the public key with the control plane without having to constantly create and publish new public keys?

I also don't understand how both WebRTC (`str0m`, right?) and Wireguard (`boringtun`) are used by connlib.

Is it that WebRTC is used as a transport layer since STUN/TURN are already standardized for hole-punching, and since it looks more like normal web traffic, and then Wireguard is layered inside for stronger and/or faster crypto than whatever WebRTC uses? (TLS?)

I think it was explained to me once but I don't know if it's documented.

File: rust/connlib/clients/shared/src/lib.rs [hunk 9]
```diff
@@ -139,27 +134,19 @@ where
...
-                runtime_stopper,
-                &callbacks
-            );
-
             // This is kinda hacky, the buffer size is 1 so that we make sure that we
             // process one message at a time, blocking if a previous message haven't been processed
             // to force queue ordering.
             let (control_plane_sender, mut control_plane_receiver) = tokio::sync::mpsc::channel(1);
 
-            let mut connection = PhoenixChannel::<_, IngressMessages, ReplyMessages, Messages>::new(Secret::new(SecureUrl::from_url(connect_url)), os_version_override, move |msg, reference, topic| {
```
Comment (ReactorScram) [inline rust/connlib/clients/shared/src/lib.rs:162]: I was a bit confused why it's `Secret<SecureUrl>` and not just `SecretUrl`, that threw me off when writing the Tauri client initially.

File: rust/connlib/shared/Cargo.toml [hunk 10]
```diff
@@ -42,6 +40,7 @@ libc = "0.2"
...
```
Comment (ReactorScram) [inline rust/connlib/shared/Cargo.toml:44]:
```suggestion
# This dependency is not ideal but will go away once we delete this implementation of the phoenix-channel client
phoenix-channel = { workspace = true }
```

What does "this implementation of the phoenix-channel client" mean? The one in connlib-shared?

File: rust/connlib/shared/src/control.rs [hunk 11]
```diff
@@ -21,32 +21,16 @@ use tokio_tungstenite::{
     tungstenite::{self, handshake::client::Request},
 };
 use tungstenite::Message;
-use url::Url;
 
 use crate::{get_user_agent, Error, Result};
+use phoenix_channel::LoginUrl;
```
Comment (ReactorScram) [inline rust/connlib/shared/src/control.rs:26]: Oh, `phoenix-channel` is part of the Firezone code. All this time I assumed it was a third-party dep.

I couldn't find a description tag, README, or module doc comment, but based on what I've heard, it is some kind of pub-sub protocol and possibly provides framing or queueing?

You said it's used inside the WebSocket connection to the API server (Or portal? Are those different), I see further down the diff that it is responsible for connecting to the control plane? So it provides pub-sub inside of a WebSocket connection? I do see it depends on `tokio-tungstenite`, a WebSocket library.

File: rust/connlib/shared/src/control.rs [hunk 12]
```diff
@@ -21,32 +21,16 @@ use tokio_tungstenite::{
...
 
 const CHANNEL_SIZE: usize = 1_000;
 const HEARTBEAT: Duration = Duration::from_secs(30);
```
Comment (ReactorScram) [inline rust/connlib/shared/src/control.rs:29]:
```suggestion
// Heartbeat sent from client, gateway, or relay to the API server (Not the same as the portal?)
const HEARTBEAT: Duration = Duration::from_secs(30);
```

I ran into this on another project - There were like 3 different heartbeats, so it was useful to just always declare the sender and receiver. In this case I don't know if the heartbeat behavior differs between:

- Endpoints (clients, gateways) talking to the control plane / API
- Clients / gateways talking to relays
- Clients talking to gateways
- Anything else

File: rust/connlib/shared/src/control.rs [hunk 13]
```diff
@@ -21,32 +21,16 @@ use tokio_tungstenite::{
...
-// See https://github.com/firezone/firezone/issues/2158
-pub struct SecureUrl {
-    pub inner: Url,
-}
-impl SecureUrl {
-    pub fn from_url(url: Url) -> Self {
-        Self { inner: url }
-    }
-}
-impl secrecy::Zeroize for SecureUrl {
```
Comment (ReactorScram) [inline rust/connlib/shared/src/control.rs:44]: Ah that's what I figured, we had to impl Zeroize to wrap `Url` in `Secret`.

So `Secret<SecureUrl>` means `Secret<ZeroizableUrl>` not `Secret<SecretUrl>`

File: rust/gui-client/src-tauri/src/client/deep_link/linux.rs (hunk 6 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/deep_link/linux.rs:15]: It should be okay.

I think Gabi had said when this was written, it's better to have it as a `Url` since it's a more specific type, and I agree in general.

But if we're treating it as opaque anyway, only parsing it into a URL once, and then re-wrapping the token in a `SecretString`, it should be fine. For deep links I don't think the URL is kept around long, it's immediately parsed and then returned as an `AuthCallback` struct or something.

File: rust/phoenix-channel/src/login_url.rs [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+// bytes)
+//
+// We are counting the nul-byte
+#[cfg(not(target_os = "windows"))]
+const HOST_NAME_MAX: usize = 256;
+
+#[derive(Clone)]
+pub struct LoginUrl {
+    url: Url,
+    host: String,
```
Comment (ReactorScram) [inline rust/phoenix-channel/src/login_url.rs:27]:
```suggestion
    // Invariant: Must stay the same as the host in `url`. This is duplicated here because `Url::host` is fallible and if we don't duplicate it, we'd have to do extra error handling in several places instead of just one place.
    host: String,
```

File: rust/phoenix-channel/src/login_url.rs [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+impl Zeroize for LoginUrl {
+    fn zeroize(&mut self) {
+        let placeholder = Url::parse("http://a.com").expect("placeholder URL to be valid");
```
Comment (ReactorScram) [inline rust/phoenix-channel/src/login_url.rs]:
```suggestion
        let placeholder = Url::parse("http://a.com").expect("placeholder URL should always be valid, it's hard-coded");
```

Might be clearer with a "should" message, otherwise the crash message will not be a complete sentence and won't add much help to the line number.

https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:7,positionColumn:1,positionLineNumber:7,selectionStartColumn:1,selectionStartLineNumber:7,startColumn:1,startLineNumber:7),source:'//+Type+your+code+here,+or+load+an+example.%0A%0A//+If+you+use+%60main()%60,+declare+it+as+%60pub%60+to+see+it+in+the+output:%0Apub+fn+main()+%7B+%0A++++None.expect(%22this+value+to+be+Some%22)%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:r1760,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'',source:1,stdinPanelShown:'1',wrap:'1'),l:'5',n:'0',o:'Executor+rustc+1.76.0+(Rust,+Editor+%231)',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

File: rust/phoenix-channel/src/login_url.rs [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+
+impl LoginUrl {
+    pub fn client<E>(
+        url: impl TryInto<Url, Error = E>,
+        firezone_token: &SecretString,
+        device_id: String,
+        device_name: Option<String>,
+        public_key: [u8; 32],
+    ) -> std::result::Result<Self, LoginUrlError<E>> {
+        let external_id = hex::encode(sha2::Sha256::digest(device_id));
```
Comment (ReactorScram) [inline rust/phoenix-channel/src/login_url.rs:48]: No action needed - In some parts of the client this was changing to "firezone-id" but I actually think "device_id" is more intuitive since it is device-specific and it's not related to the actor or anything.

I don't remember why we were changing it to "firezone-id"

File: rust/phoenix-channel/src/login_url.rs [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+        &self.url
+    }
+
+    // TODO: Only temporarily public until we delete other phoenix-channel impl.
+    pub fn host(&self) -> &str {
+        &self.host
+    }
+}
+
+fn parse_host<E>(url: &Url) -> Result<String, LoginUrlError<E>> {
```
Comment (ReactorScram) [inline rust/phoenix-channel/src/login_url.rs:135]:
```suggestion
/// Host, including port if present. e.g. `example.com:8080`
fn parse_host<E>(url: &Url) -> Result<String, LoginUrlError<E>> {
```

File: rust/phoenix-channel/src/login_url.rs [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+    InvalidUrl(E),
+    #[error("the url is missing a host")]
+    MissingHost,
+}
+
+#[cfg(not(target_os = "windows"))]
+fn get_host_name() -> Option<String> {
+    let mut buf = [0; HOST_NAME_MAX];
+    // SAFETY: we just allocated a buffer with that size
+    if unsafe { libc::gethostname(buf.as_mut_ptr() as *mut _, HOST_NAME_MAX) } != 0 {
```
Comment (ReactorScram) [inline rust/phoenix-channel/src/login_url.rs:158]: What was the reason to not use the `hostname` crate on all platforms? internally they make the same call https://github.com/svartalf/hostname/blob/master/src/nix.rs#L25

Although, they don't enforce the 256 byte limit that we do. They allow higher if the OS does, it looks like.

**Review by ReactorScram — APPROVED** [review]
LGTM but I'm asking a lot of questions again so I can get to understand the connlib architecture.

File: rust/connlib/clients/android/src/lib.rs (hunk 2 above)
Comment (conectado) [reply rust/connlib/clients/android/src/lib.rs:425]: > I like that this opens a possibility for clients to persist their private key to disk if we ever want TOFU semantics with an even stronger "We can't decrypt your traffic even if we want" claim after 1.0

Just as a note, I don't think there's a stronger claim to privacy in this case, since we already generate the key in-memory, if we wanted to maliciously leak the key we could do it either way by sending the private key to the control plane.

File: rust/connlib/clients/shared/src/lib.rs (hunk 8 above)
Comment (conectado) [reply rust/connlib/clients/shared/src/lib.rs:7]: > I saw the Dalek docs said it's better to use EphemeralSecret. I assume the reason we use StaticSecret is so we can share the public key with the control plane without having to constantly create and publish new public keys?

We don't use the `StaticSecret` to generate a DH `SharedSecret` but we use the `StaticSecret` both to generate the public key, which we indeed only use once and the bytes to encrypt messages with Wireguard, we can only obtain the underlying bytes with a `StaticSecret`.

Comment (conectado) [reply rust/connlib/clients/shared/src/lib.rs:7]: > I also don't understand how both WebRTC (str0m, right?) and Wireguard (boringtun) are used by connlib.

> Is it that WebRTC is used as a transport layer since STUN/TURN are already standardized for hole-punching, and since it looks more like normal web traffic, and then Wireguard is layered inside for stronger and/or faster crypto than whatever WebRTC uses? (TLS?)

we no longer use WebRTC, we use only the ICE part of `str0m`, which provides us with the transport(hole-punching and all that). `Wireguard` is used to create a tunnel over that, basically encrypting/decrpyting the packets that are sent through the transport. We could use DTLS like webrtc does but it is more complex than Wireguard and we don't need that complexity.

File: rust/phoenix-channel/src/login_url.rs (hunk 18 above)
Comment (thomaseizinger) [reply rust/phoenix-channel/src/login_url.rs:158]: I don't know, I am just moving code around here :)

File: rust/connlib/clients/shared/src/lib.rs (hunk 8 above)
Comment (thomaseizinger) [reply rust/connlib/clients/shared/src/lib.rs:7]: > I also don't understand how both WebRTC (`str0m`, right?) and Wireguard (`boringtun`) are used by connlib.
> 
> Is it that WebRTC is used as a transport layer since STUN/TURN are already standardized for hole-punching, and since it looks more like normal web traffic, and then Wireguard is layered inside for stronger and/or faster crypto than whatever WebRTC uses? (TLS?)

We don't use the full webrtc stack, just ICE to figure out a UDP "connection" (two sockets, unencrypted, unauthenticated). Then we layer wireguard on top of that. That is the scope of `snownet`.

The tunnel then adds ACLs to resources on top of that and actually pushes the bytes between device and sockets around :)

**Comment by thomaseizinger** [comment]
> LGTM but I'm asking a lot of questions again so I can get to understand the connlib architecture.

Please do! :)

File: rust/connlib/shared/src/control.rs (hunk 11 above)
Comment (thomaseizinger) [reply rust/connlib/shared/src/control.rs:26]: Our control plane and portal are both built in Elixir. The portal uses Phoenix LiveView for real time updates of the UI. Phoenix LiveView is built on top of a generic pub-sub system that works with WebSockets.

The `phoenix-channel` Rust crate is our implementation of a client for the same pubsub system. It is used by gateways, relays and soon the clients.

We have two implementations because Gabi wrote one for connlib and I wrote one for the relay.

**Review by conectado — APPROVED** [review]

File: rust/connlib/shared/src/control.rs (hunk 12 above)
Comment (thomaseizinger) [reply rust/connlib/shared/src/control.rs:29]: This is just clients to portal but will soon be deleted.

File: rust/connlib/shared/Cargo.toml (hunk 10 above)
Comment (thomaseizinger) [reply rust/connlib/shared/Cargo.toml:44]: Yes.

File: rust/connlib/clients/android/src/lib.rs (hunk 2 above)
Comment (ReactorScram) [reply rust/connlib/clients/android/src/lib.rs:425]: Well it's not about leaking but MITMing. If the control plane lies to the client and gateway and instead of giving them each other's pubkeys, it gives them the pubkeys of a relay it controls, and also passes off the relay as STUN/TURN candidate, the relay could decrypt traffic.

AIUI if the end user does not take any responsibility for key management, some kind of MITM attack is always possible. Otherwise they are ultimately trusting someone else to not lie about the pubkey.

Comment (conectado) [reply rust/connlib/clients/android/src/lib.rs:425]: I mean if we get the private key we can just decrypt all traffic, whether we get the key from disk or create it in memory doesn't change that.

Comment (ReactorScram) [reply rust/connlib/clients/android/src/lib.rs:425]: I mean if the client relies on the API / portal to learn the gateway's pubkey, the portal could lie about the pubkey and MITM the client's traffic. Right?

You'd have to either have persistent keys and then you can do TOFU, or do a CA.

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs:425]: > and MITM the client's traffic

The portal is not in the data plane, so it doesn't see any traffic.

Relays are, and we already have a couple options for handling the MITM possibility for the truly paranoid there that we could roll out.

Comment (ReactorScram) [reply rust/connlib/clients/android/src/lib.rs:425]: Yeah the portal would have to direct traffic to a relay that was also in on this

Comment (thomaseizinger) [reply rust/connlib/clients/android/src/lib.rs:425]: > Yeah the portal would have to direct traffic to a relay that was also in on this

That is not up to the portal to decide though, we will form a direct connection if we can, even if we are given a relay from the portal.

Comment (ReactorScram) [reply rust/connlib/clients/android/src/lib.rs:425]: But if the client* doesn't know the IP:port of the gateway, can the control plane not say, "Hm you're both behind unbreakable NATs, use this IP:port combo (which I secretly control) and here is the pubkey for your peer (which I secretly generated), trust me"?

\* And vice verse for gateway not knowing where the client is

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs:425]: @ReactorScram Let's continue this discussion offline.
